roles/haproxy/tasks/get_acme_certs.yml

```diff
@@ -1,34 +1,98 @@
 ---
-- name: 
+- name: Create list of domains to request certificates for
   ansible.builtin.set_fact:
-    haproxy_ssl_hosts: "{{ (((haproxy_applications | map(attribute='vhost_name') | list) + (haproxy_redirects | map(attribute='hostname') | list)) | unique) | join (' -d ') }}"
+    haproxy_ssl_hosts: >
+      {{
+        (
+          [base_domain]
+          + (haproxy_applications | map(attribute='vhost_name') | list)
+          + (haproxy_redirects | map(attribute='hostname') | list)
+        )
+        | unique | sort
+      }}
 
-- name: Install the certificates using acme
-  ansible.builtin.shell:
-    cmd: "/home/acme/.acme.sh/acme.sh --issue --stateless --keylength 2048  -d {{ base_domain }} -d {{ haproxy_ssl_hosts }} --server https://acme.sectigo.com/v2/GEANTOV"
-  become_user: acme
+- name: Check existence of acme CNAME records
+  ansible.builtin.command:
+    cmd: "dig +short -t CNAME '_acme-challenge.{{ item }}'"
+  register: haproxy_acme_cname
+  failed_when: "haproxy_acme_cname.stdout == ''"
+  changed_when: false
+  loop: "{{ haproxy_ssl_hosts }}"
+  delegate_to: "localhost"
+  run_once: true
+
+- name: Issue the certificates using acme
+  ansible.builtin.command:
+    cmd: |
+      /home/acme/.acme.sh/acme.sh
+        --issue
+        --keylength 2048
+        --days "{{ haproxy_acme_renewal_days }}"
+        --dns dns_acmedns
+        --stateless
+        --dnssleep 3
+        --server "{{ haproxy_acme_server }}"
+        --domain "{{ item }}"
+  environment:
+    ACMEDNS_BASE_URL: "{{ haproxy_acmedns.baseurl }}"
+    ACMEDNS_USERNAME: "{{ haproxy_acmedns.username }}"
+    ACMEDNS_PASSWORD: "{{ haproxy_acmedns.password }}"
+    ACMEDNS_SUBDOMAIN: "{{ haproxy_acmedns.subdomain }}"
+  loop: "{{ haproxy_ssl_hosts }}"
+  register: "acme_issue"
+  changed_when: "acme_issue.rc == 0"
+  failed_when: 'acme_issue.rc != 0 and acme_issue.rc != 2'
+  become_user: "acme"
   become: true
 
-- name: Deploy the certificates to haproxy without the need for haproxy to run
-  ansible.builtin.shell:
-    cmd: "DEPLOY_HAPROXY_PEM_PATH=/etc/haproxy/certs   /home/acme/.acme.sh/acme.sh --deploy -d {{ base_domain }} --deploy-hook haproxy"
-  become_user: acme
+# Now deploy the certificates to haproxy
+# We do this in two steps:
+# 1. Cold-deploy the certificates to haproxy's certs directory
+# 2. Make sure haproxy is started, and hot-deploy the certificates to haproxy though the socket API
+
+- name: Cold-deploy the certificates to haproxy  # noqa: no-handler
```

Comment:
What do you mean by "noqa: no-handler"?

Reply:
This is to satisfy ansible-lint, which doesn't like `when: foo.changes` tasks.  Generally, those should probably be changd to handlers, but in this case that would complicate the role quite a bit.